### PR TITLE
Assignment5

### PR DESCRIPTION
**Pull request by Zophed**

### CODE REVIEW

**Comment by worseTyler** [comment]
I will review this

File: GenericsHomework/GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                _Value = value ?? throw new ArgumentNullException(nameof(value));
+            }
+        }
+        public Node<T> Next
+        {
+            get => _Next!;
+
+            private set
+            {
+                value._Next = this;
```
Comment (worseTyler) [inline GenericsHomework/GenericsHomework/Node.cs:25]: I don't think you need this line

File: GenericsHomework/GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+                _Next = value ?? throw new ArgumentNullException(nameof(value));
+            }
+        }
+
+        public Node(T value)
+        {
+            this.Value = value;
+            this.Next = this;
```
Comment (worseTyler) [inline GenericsHomework/GenericsHomework/Node.cs:33]: Remove the this prefix, also on line 39

File: GenericsHomework/GenericsHomework/Node.cs (hunk 1 above)
Comment (worseTyler) [inline GenericsHomework/GenericsHomework/Node.cs:25]: This makes the list circular, gotcha

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+            //Act
+
+            //Assert
+            Assert.AreEqual<string>("42", intNode.ToString());
+            Assert.AreEqual<string>("The Dread Pirate Roberts", stringNode.ToString());
+        }
+
+        [TestMethod]
+        public void NextProperty_ReturnsCorrectNode()
```
Comment (worseTyler) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:41]: Does this test the next property?

**Comment by worseTyler** [comment]
Instructions
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
Include a constuctor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ✔
In addition to non-null validation, the body of the Next property should insert the next Node (the value) into the list. ✔
Add an Insert method that takes a value and inserts a new Node instance after the current node (by invoking the Next property). ✔
Add a clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage colleciton because all the items point to each other and therefore may never be garbage collected. ✔
You should not rely on any BCL generic classes for your implementation. ✔
Extra Credit
Do one of the following two options (or both if you want extra extra credit) :)

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌
Each circle contains n items and each item can belong to one more more Circle instances.
You are not required to use a Node from earlier in the homework for your venn diagram implementation.
You are welcome to use exising BCL generic classes for the extra credit.
Implement Systm.Collections.Generic.ICollection<T> on the Node class ❌
Fundamentals
Ensure you enable nullable reference types, net5 targetted, C# 9.0, warnings as errors, and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEquals<T>() (the generic version)
All of the above should be unit tested ✔l

**Comment by mmwoodfo** [comment]
When doing your code reviews it is more beneficial for the person you are reviewing to give them inline comments (similar to what @worseTyler did on this PR) rather than a single comment listing all of the issues. This isn't something you are going to be dinged for on this assignment, but you may lose points on future assignments for unclear or difficult to read reviews.'

It's fine to keep the large block comment with the full checklist of items done well and not done well, but in the future try to give inline comments for any 'X's that you leave on the review rather than adding the reasoning to the block of text.

File: GenericsHomework/GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
+﻿using System;
+
+namespace GenericsHomework
+{
+    public class Node<T>
+    {
+        private T? _Value;
+        private Node<T>? _Next;
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework/Node.cs:8]: If you set these in the constructor you do not have to make them as nullable.

File: GenericsHomework/GenericsHomework/Node.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            this.Next = this;
+        }
+
+        public override string ToString()
+        {
+            if (Value is null)
+            {
+                throw new ArgumentNullException(nameof(Value));
+            }
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework/Node.cs:58]: The code below will return an empty string if `Value` is null, so you don't need this here.

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            //Act
+            stringNode.Clear();
+
+            //Assert
+            Assert.IsTrue(stringNode.ToString() == stringNode.Next.ToString());
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void ToString_PassedNullNode_ThrowsException()
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:83]: This test name is misleading.

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            //Assign
+            Node<int> headNode = new Node<int>(42);
+
+            //Act
+            headNode.Insert(43);
+
+            //Assert
+            Assert.AreEqual<string>("42", headNode.ToString());
+            Assert.AreEqual<string>("43", headNode.Next.ToString());
+            Assert.AreEqual<string>("42", headNode.Next.Next.ToString());
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:65]: Try to avoid using your `ToString` method when unit testing unrelated code. Your unit tests should be focused on a single _unit_ of your code, in this case, just the `Insert` method. Another way to do this below:
```suggestion
            Node<int> secondNode = headNode.Next;

            Assert.AreEqual(secondNode, headNode.Next);
            Assert.AreNotEqual(secondNode, headNode);
            Assert.AreEqual(headNode, headNode.Next.Next);
```

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        public void ToString_ReturnsCorrectString()
+        {
+            //Assign
+            var (stringNode, intNode) = TestNodes();
+
+            //Act
+
+            //Assert
+            Assert.AreEqual<string>("42", intNode.ToString());
+            Assert.AreEqual<string>("The Dread Pirate Roberts", stringNode.ToString());
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:37]:
```suggestion
            Assert.AreEqual("42", intNode.ToString());
            Assert.AreEqual("The Dread Pirate Roberts", stringNode.ToString());
```

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            Node<string> secondNode = new Node<string>("Princess Buttercup");
+
+            //Act
+
+            //Assert
+            Assert.AreNotEqual<Node<string>>(firstNode, secondNode);
+        }
+
+        [TestMethod]
+        public void InsertMethodGivenNewNode_AssignsNextAndLoopsBackToNode()
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:54]: You should test more than two nodes being inserted into the list. You should also test the order in which the nodes get added to the list. When writing unit tests the goal is to write unit tests that stress test your code. Hit the limits, test edge cases, and hopefully reveal bugs.

File: GenericsHomework/GenericsHomework/Node.cs (hunk 1 above)
Comment (mmwoodfo) [reply GenericsHomework/GenericsHomework/Node.cs:25]: @worseTyler @Zophed
Not quite. While this works with two nodes in the list, once you add three you will start to get _detached head and neck_ like structure in your list.

The following unit test will cause a failure in your code.
```C#
// 1 -> 2 -> 3
// ┖────┘

// Arrange
Node<int> one = new Node<int>(1);

// Act
one.Next.Insert(2);
var two = one.Next;

two.Next.Insert(3);
var three = two.Next;

// Assert (These will fail with your code)
Assert.AreEqual(one, two.Next.Next);
Assert.AreEqual(one.Next.Next, three);
Assert.AreEqual(three, two.Next);
```

Instead of a linked list that looks like this:
```
1 -> 2 -> 3
┖─────────┘
```

You'll have this:
```
1 -> 3
┖────┘

2 -> 1 -> 3
     ┖────┘
```

I see you made a comment about this yesterday - so be sure to add robust tests in the future.

There are easy ways within visual studio to look at your variables and data structures. [Here are some tips for that](https://devblogs.microsoft.com/devops/7-ways-to-look-at-the-values-of-variables-while-debugging-in-visual-studio/).

File: GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            headNode.Insert(43);
+
+            //Assert
+            Assert.AreEqual<string>("42", headNode.ToString());
+            Assert.AreEqual<string>("43", headNode.Next.ToString());
+            Assert.AreEqual<string>("42", headNode.Next.Next.ToString());
+        }
+
+        [TestMethod]
+        public void ClearMethodRan_AllNodesExceptCurrentRemoved()
```
Comment (mmwoodfo) [inline GenericsHomework/GenericsHomework.Tests/NodeTests.cs:69]: You could also add a test to make sure the detached nodes get collected by the garbage collector.
